### PR TITLE
Standalone sidecar charm

### PR DESCRIPTION
**Pull request by misohu**
Deploy MLflow as a sidecar charm (without any relation).

MLflow is running and is exposed as NodePort on port speciffied in config. 

This is a follow up PR on https://github.com/canonical/mlflow-operator/pull/103 (first review the 103).

**Publish charm will fail as it is merging agains feature branch not main (but it works on main check this [draft](https://github.com/canonical/mlflow-operator/pull/101) pr against main with the same code)**

### CODE REVIEW

File: metadata.yaml
Comment (DnPlas) [inline metadata.yaml:15]: Should we use a tag here? Maybe we can keep a 1:1 relation with upstream in terms of image tags, in the end this image corresponds to a certain mlflow version, no?

**Review by DnPlas — COMMENTED** [review]
Thansk @misohu, I reviewed commits 9d03e6228fd2bc20f1b2b4d4f857d0983ebbe190 and e9ee80ee2b4dff06b696539cf24b17a941514669, just left a comment.

File: metadata.yaml
Comment (misohu) [reply metadata.yaml:15]: for now we have just latest as we need to adjust the image's CI .... we cant use upstream tag as we may want to have multiple versions of upstream tag in case we will fix vulnerabilities. There is a task in jira for that for now we have just latest tag.

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:18]:
```suggestion
    """A Juju Charm for MLFlow."""
```

Comment (ca-scribner) [reply src/charm.py:18]: Can you check your linting config vs our other charms?  I'd have expected this to fail because it doesn't end in ".", so I think the config is not the same as our other charms

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:41]: Good to know, but probably not needed :)

```suggestion
```

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:40]: This if/else might be more readable if it only set the stuff that differs between the two branches, rather than have it all duplicated.  I think atm it is somehing like:
```python
if True:
  x = "true branch"
  y = "true branch"
  do_something(
    common_setting,
    x,
    y
  )
else:
  x = "false branch"
  y = "false branch"
  do_something(
    common_setting,
    x,
    y
  )
```
which makes the reader need to focus to understand what is different.  What about writing it like:
```python
if True:
  x = "true branch"
  y = "true branch"
else:
  x = "false branch"
  y = "false branch"
do_something(
  common_setting,
  x,
  y
)
```
that way the common parts are easy to spot

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:110]: Should this be an Error rather than just something Blocked?  Blocked suggests it is something that a user can intervene about (change a setting, etc), but I don't think that's the case here

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:103]: Should this be an Error rather than just something Blocked? Blocked suggests it is something that a user can intervene about (change a setting, etc), but I don't think that's the case here.

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:121]: I think this log message is too aggressive.  We're not failing to handle errors here, we get here by *handling* errors.  For example, if this unit is not the leader it is going to say "Failed to handle X with error: 'Not a leader, skipping setup'" and go into `WaitingStatus`.  I like that we log the message, but not that we call it an unhandled error
```suggestion
            self.logger.info(f"Event {event} stopped early with message: {str(err)}")
```

File: tests/unit/test_operator.py
Comment (ca-scribner) [inline tests/unit/test_operator.py:77]: Is there a better name we can give this?  It tests more than just "update_layer_success"

**Review by ca-scribner — COMMENTED** [review]
minor stuff, otherwise lgtm

File: src/charm.py
Comment (misohu) [reply src/charm.py:110]: Blocked is consistent with rest of our rewrites .... do we want to change it in every sidecar charm? If so Lets keep this Blocked and create a task. What do you think?

File: src/charm.py
Comment (misohu) [reply src/charm.py:103]: Blocked is consistent with rest of our rewrites .... do we want to change it in every sidecar charm? If so Lets keep this Blocked and create a task. What do you think?

File: src/charm.py
Comment (misohu) [reply src/charm.py:41]: Solved thanks :)

File: src/charm.py
Comment (DnPlas) [reply src/charm.py:105]: I agree, a `BlockedStatus` means the charm needs a human operator to unblock, a `WaitingStatus` could be more appropriate here. I think this should be fixed before we merge your changes from the dev branch into main, but this PR is already very complex and convoluted. I am okay if you send a follow up PR just for this.

File: src/charm.py
Comment (DnPlas) [reply src/charm.py:98]: I agree, this should get the unit in `Error` state, let's work on it in a separate PR before merging all the changes from dev branch into main.

**Review by DnPlas — APPROVED** [review]
Thanks @misohu, I think there are a few comments that we can resolve in a separate PR as this one is already very complex and convoluted. Please make sure to resolve the conversations of the status before we merge your dev branch into main.

**Pull request merged**
